### DISCUSSION TITLE
Difference between data provided to handleSubmit and getValues()

### DISCUSSION THREAD

**Discussion by kwassholm-work** [react-hook-form/react-hook-form#6142] [Q&A]
Heya,

I'm seeing different data in handleSubmit's form data vs. getValues() called inside the same handleSubmit.
I'm haven't been able to create a simplified reproducible case yet, so unfortunately I don't have a codesandbox to give out yet.
Hence I would like to first confirm my understanding that should these two have the same data, i.e. there's a bug somewhere, or is there something I'm missing here?

My particular use case is a form (with shouldUnregister set to true) where the user can dynamically enable/disable nested text-field inputs.
If the user disables one or more of the nested text-field inputs the component is unmounted (I can verify the unmounting happens with React DevTools).
But the data from the unmounted text-field is still provided to the handleSubmit.
But getValues() does not contain the unmounted component's data.
I would assume that the behaviour with getValues() should be the correct one?

**Comment by bluebill1049** [comment]
GetValues will merge default values, handleSubmit don't.

**Reply by kwassholm-work** [reply]
OK, that I understood from the API docs, but I don't understand how that works with a form that has "shouldUnregister=true. That is, why does handleSubmit still have unmounted fields in its data but getValues() does not?
Shouldn't handleSubmit also have data where unmounted fields are removed?

**Reply by bluebill1049** [reply]
shouldUnregister: true  means inputs will gets removed after unmount

**Reply by kwassholm-work** [reply]
Yes, but why that is not reflected in the handleSubmit? shouldUnregister: true seems to only work with getValues in my case.

**Reply by bluebill1049** [reply]
can you provide a simple example?

**Reply by bluebill1049** [reply]
codesandbox

**Reply by kwassholm-work** [reply]
Yes, I'll try to provide a sample case.

**Reply by kwassholm-work** [reply]
https://codesandbox.io/s/remote-handlesubmit-forked-f5gk4?file=/src/App.js

Here's the simplified version of our use case. For some reason in this example both the getValues() and handleSubmit behave the same way; both have input values from unmounted inputs. Whereas in our own code only the handleSubmit was behaving that way.

Anyways, the issue is the same. I would assume that after you enable one or more checkboxes and then disable them, then whenever you submit you would only get the values from the inputs that are still mounted. Or is there something wrong in my example on how I'm using the library?

**Reply by kwassholm-work** [reply]
Any idea if there's a bug or have I misunderstood how it's supposed to work and/or missing something?

**Reply by bluebill1049** [reply]
you should build a controlled component to host the data in there (checkboxes) and wrapped with Controller for hook form to collect and validate data.

**Reply by kwassholm-work** [reply]
Ok, thanks, will do so.

But I still think the API docs is a bit misleading in https://react-hook-form.com/api/useform where it says about "shouldUnregister=true":

> 
> set shouldUnregister: true will set your form behave more closer as native.
>     - input unmount will remove value.
>     - input hidden should applied for hidden data.
>     - only registered input will be included as submission data.

Especially the last point; the unmounted and unregistered input is still present in the submission data.

**Reply by bluebill1049** [reply]
if you have 3 checkboxes with the same name, hook form will only remove value until all checkboxes are removed. it's group units.

**Reply by kwassholm-work** [reply]
The problem is not with the checkboxes. The checkboxes only control the visibility of the text input fields, like it is in codesandbox. And the text inputs have unique names.

`name={`lastName.${i}.name`}`

I don't see "_only registered input will be included as submission data_" working as specified. Only way I see submitted data not containing the unmounted fields is if I manually call unregister() for the unmounted text input fields.

**Reply by bluebill1049** [reply]
<img width="847" alt="Screen Shot 2021-08-13 at 7 40 22 pm" src="https://user-images.githubusercontent.com/10513364/129337934-3aaf94b6-0ed6-4be8-ad8b-32fa824d975f.png">

look at your example, your example is `register` with the same **name**. I suggest you try out with a single input with unique name and unmount that input and check out the submit value.

**Reply by kwassholm-work** [reply]
If you log that "name" variable you should see:

```
lastName.0.name 
lastName.1.name 
lastName.2.name 
```

That name comes from `name={`lastName.${i}.name`}`.
Please note that there two input's inside the React component; one which is a checkbox (which does have the same name as other instances of the checbkoxes as intended), and one input which has unique names as listed above.

**Reply by kwassholm-work** [reply]
https://codesandbox.io/s/remote-handlesubmit-forked-lse77?file=/src/App.js

Here is an example with just one and not nested input field. It seems to exhibit the same behaviour; field is not unregistered after unmounting. Only after I manually unregister it will disappear from the submission data.

**Reply by bluebill1049** [reply]
https://codesandbox.io/s/musing-pine-w60o1?file=/src/App.js

```tsx
import * as React from "react";
import { useForm } from "react-hook-form";
import Headers from "./Header";
import "./styles.css";

let renderCount = 0;

export default function App() {
  const [value, setValue] = React.useState(0);
  const { register, handleSubmit } = useForm({
    shouldUnregister: true
  });
  const onSubmit = (data) => console.log(data);
  renderCount++;

  return (
    <div>
      <Headers
        renderCount={renderCount}
        description="Performant, flexible and extensible forms with easy-to-use validation."
      />

      <form onSubmit={handleSubmit(onSubmit)}>
        <input onChange={(e) => setValue(e.target.value)} type="number" />
        {[...Array(+value).keys()].map((value, i) => (
          <input
            key={i}
            value={value}
            type="checkbox"
            {...register(`lastName.${i}.name`)}
            placeholder="First Name"
          />
        ))}
        <input type="submit" />
      </form>
    </div>
  );
}
```

take a look at this example above, by changing the input number.

**Reply by kwassholm-work** [reply]
Please try the codesandboxes I provided. It's not about the checkboxes but the text input fields.
Even in my example the checkboxes do get unregistered and not submitted but not the text input fields, as I would assume.

**Reply by bluebill1049** [reply]
the problem is due to the `useEffect` at `useForm` never gets executed, because state update is isolated in useWatch level, will improve on this problem in the 7.13.0 release (most likely next week)

**Reply by kwassholm-work** [reply]
Great, thanks! I'll test it again when it is released.
